### PR TITLE
Defining steps without echo

### PR DESCRIPTION
**Pull request by kevgo**
@charlierudolph @allewun @ricmatsui 

We run more and more into the situation where commands have to interact with the user in some form to determine the correct set of steps. An example is #663.

The current architecture uses `echo` to gather all the commands to be run by a GT command. Echo is overused in Bash, and using it for data flow here prevents us from using it to print information to the user during the step determination workflow. In the end, all we do with these steps is to write them to a file.

This PR refactors the steps infrastructure. It adds a dedicated `step` command, which appends the given command to $STEPS_FILE directly. This enables us to use the `echo` command in the step determination workflows down the line. It also makes step definitions blocks in the various GT commands more readable.

### CODE REVIEW

**Comment by allewun** [comment]
I like it :ship:

File: src/helpers/script_helpers.sh [hunk 1]
```diff
@@ -191,6 +191,41 @@ function skippable {
...
+  fi
+}
+
+
+# Adds the given step if all the given conditions are true
+function step_if_all_true {
+  local command="$1"
+  shift
+
+  local runStep=true
```
Comment (charlierudolph) [inline src/helpers/script_helpers.sh]: I think `addStep` is more accurate

File: src/helpers/script_helpers.sh [hunk 2]
```diff
@@ -191,6 +191,41 @@ function skippable {
 }
 
 
+# Adds a step to the current command
+function step {
```
Comment (charlierudolph) [inline src/helpers/script_helpers.sh]: How about naming this `add_step`. `step` doesn't say enough to me about what the function does

File: src/git-extract [hunk 3]
```diff
@@ -58,9 +58,9 @@ function preconditions {
 
 function steps {
```
Comment (charlierudolph) [inline src/git-extract]: What do you think about renaming these functions to add_steps?

Comment (kevgo) [inline src/git-extract]: Hmm, what this method does is defining the steps that this command executes. `define_steps` is a bit long. So calling it `steps`, meaning this method defines and returns the steps that this command runs, seems appropriate here. 

If there is a specifier like `add` in `add_steps`, it feels like there should also be a `remove_steps`, a `run_steps` etc in this API.

Comment (charlierudolph) [inline src/git-extract]: I don't think `define_steps` is long. I would be good with that

File: src/helpers/script_helpers.sh (hunk 2 above)
Comment (kevgo) [inline src/helpers/script_helpers.sh]: By itself I would agree with you. Given that it is called in a method called `steps` that defines the steps for a command, I think its okay and nicely short. I don't feel strongly about this though, so am happy to adjust.

File: src/helpers/script_helpers.sh (hunk 1 above)
Comment (charlierudolph) [inline src/helpers/script_helpers.sh]: Thinking about it more how about `allTrue`?

**Comment by charlierudolph** [comment]
Would like the rename from runStep to allTrue in script helpers. Otherwise good to :ship:

**Pull request closed without merging**